drivers/adc/adc_sam0.c

```diff
@@ -231,20 +231,27 @@ static int adc_sam0_channel_setup(const struct device *dev,
 	if (channel_cfg->differential) {
 		INPUTCTRL |= ADC_INPUTCTRL_MUXNEG(channel_cfg->input_negative);
 
-#ifdef ADC_INPUTCTRL_DIFFMODE
+#if defined(ADC_INPUTCTRL_DIFFMODE)
 		INPUTCTRL |= ADC_INPUTCTRL_DIFFMODE;
-#else
+#elif defined(ADC_CTRLB_DIFFMODE)
 		adc->CTRLB.bit.DIFFMODE = 1;
-		wait_synchronization(adc);
+#elif defined(ADC_CTRLC_DIFFMODE)
+		adc->CTRLC.bit.DIFFMODE = 1;
+#else
+		return -EINVAL;
 #endif
 	} else {
 		INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;
 
-#ifndef ADC_INPUTCTRL_DIFFMODE
+#if defined(ADC_INPUTCTRL_DIFFMODE)
+		/* no op */
```

Comment:
Why?

Reply:
Why the `/* no op */`?

Because this makes it explicit that there are three paths available here (not just two), and also maintains symmetry with the other side of the `if {} else {}`.

Can drop it if you want.

Reply:
Driver set this:
`INPUTCTRL |= ADC_INPUTCTRL_DIFFMODE;`
then
`INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;`

So, when driver set `ADC_INPUTCTRL_MUXNEG_GND` it automatically disable `ADC_INPUTCTRL_DIFFMODE` ?

Reply:
> So, when driver set ADC_INPUTCTRL_MUXNEG_GND it automatically disable ADC_INPUTCTRL_DIFFMODE ?

By virtue of the `int INPUTCTRL = 0` above, yes...?

This "_collected_" value is finally written to the register here:

https://github.com/zephyrproject-rtos/zephyr/blob/0d412c4ce75c2f2ae0a6eb538627bbd9b04278ed/drivers/adc/adc_sam0.c#L235

---

`DIFFMODE` and `DSEQSTOP` fields are not present in this register for the SAML21 or SAMD21(resolved in `fixup_adc.h`)... snippet below is from SAMD5x.

![image](https://user-images.githubusercontent.com/36298367/146964413-91791745-518e-4a75-881a-3f000ef103d0.png)

Reply:
@nandojve please resolve if you're happy, or let me know if not?